accessor/reference_helper.hpp

```diff
@@ -0,0 +1,188 @@
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_ACCESSOR_REFERENCE_HELPER_HPP_
+#define GKO_ACCESSOR_REFERENCE_HELPER_HPP_
+
+
+#include <type_traits>
+#include <utility>
+
+
+#include "utils.hpp"
+
+
+// CUDA TOOLKIT < 11 does not support constexpr in combination with
+// thrust::complex, which is why constexpr is only present in later versions
+#if defined(__CUDA_ARCH__) && defined(__CUDACC_VER_MAJOR__) && \
+    (__CUDACC_VER_MAJOR__ < 11)
+
+#define GKO_ACC_ENABLE_REFERENCE_CONSTEXPR
+
+#else
+
+#define GKO_ACC_ENABLE_REFERENCE_CONSTEXPR constexpr
+
+#endif  // __CUDA_ARCH__ && __CUDACC_VER_MAJOR__ && __CUDACC_VER_MAJOR__ < 11
+
+
+namespace gko {
+namespace acc {
+/**
+ * This namespace is not part of the public interface and can change without
+ * notice.
+ */
+namespace detail {
+
+
+/**
+ * This is a mixin which defines the binary operators for *, /, +, - for the
+ * Reference class, the unary operator -, and the assignment operators
+ * *=, /=, +=, -=
+ * All assignment operators expect an rvalue reference (Reference &&) for
+ * the Reference class in order to prevent copying the Reference object.
+ *
+ * @tparam Reference  The reference class this mixin provides operator overloads
+ *                    for. The reference class must overload the cast
+ *                    operator to ArithmeticType
+ *
+ * @tparam ArithmeticType  arithmetic type the Reference class is supposed
+ *         to represent.
+ *
+ * @warning  This struct should only be used by reference classes.
+ */
+template <typename Reference, typename ArithmeticType>
+struct enable_reference_operators {
+    using arithmetic_type = std::remove_cv_t<ArithmeticType>;
+
+    constexpr GKO_ACC_ATTRIBUTES GKO_ACC_INLINE arithmetic_type
+    to_arithmetic_type() const
+    {
+        return static_cast<const Reference*>(this)
+            ->Reference::operator arithmetic_type();
+    }
```

Comment:
This function calls the cast operator (`operator arithmetic_type()`) of the `this` object.
Because this cast operator is defined in the `Reference` class, It needs to be cast to `Reference` first, before the operator is called. Since this is a Mixin and `Reference` is supposed to inherit from it, this cast should always be succesfull (also, I assume if it would not be inherited, this should yield in a compiler error).

In summary, this has 2 layers:
1. This calls the cast operator to `arithmetic_type` that must be present (otherwise, a proxy reference would not make sense). If it is not present, it will result in a compilation error.
2. Since `enable_reference_operators` does not have that cast operator, it needs to be cast to `Reference` first before the operator can be found.